lib/sparrow/telemetry/timer.ex

```diff
@@ -0,0 +1,85 @@
+defmodule Sparrow.Telemetry.Timer do
+  @moduledoc """
+  Module responsible for handling emitting telemetry events which measure time of execution
+  """
+  require Logger
+
+  defmacro __using__(_mod) do
+    quote do
+      import Sparrow.Telemetry.Timer
+      Module.register_attribute(__MODULE__, :timed_functions, accumulate: true)
+      @on_definition Sparrow.Telemetry.Timer
+      @before_compile Sparrow.Telemetry.Timer
+    end
+  end
+
+  def __on_definition__(env, _kind, name, args, guards, body) do
+    module = env.module
+    timed = Module.get_attribute(module, :timed)
+
+    if timed do
+      Module.put_attribute(module, :timed_functions, %{
+        name: name,
+        args: args,
+        guards: guards,
+        body: body
+      })
+
+      Module.delete_attribute(module, :timed)
+    end
+  end
+
+  defmacro __before_compile__(env) do
+    module = env.module
+    timed_funs = Module.get_attribute(module, :timed_functions)
+
+    new_funs =
+      timed_funs
+      |> Enum.map(fn fun_info ->
+        :ok =
+          Module.make_overridable(module, [
+            {fun_info.name, length(fun_info.args)}
+          ])
+
+        new_body = update_body(fun_info)
+
+        if length(fun_info.guards) > 0 do
+          quote generated: true do
+            def unquote(fun_info.name)(unquote_splicing(fun_info.args))
+                when unquote_splicing(fun_info.guards) do
+              unquote(new_body)
+            end
+          end
+        else
+          quote generated: true do
+            def unquote(fun_info.name)(unquote_splicing(fun_info.args)) do
+              unquote(new_body)
+            end
+          end
+        end
+      end)
+
+    quote do
+      (unquote_splicing(new_funs))
+    end
+  end
+
+  defp update_body(fun_info) do
+    quote do
+      t_start = Time.utc_now()
+      res = unquote(Keyword.get(fun_info.body, :do))
+      t_end = Time.utc_now()
+      t = abs(Time.diff(t_start, t_end, :microsecond))
+
+      :telemetry.execute(
+        [:sparrow, __MODULE__, unquote(fun_info.name)],
```

Comment:
All other, manual events have a different names (`:h2_worker` vs `Elixir.Sparrow.H2Worker`). We probably should go with explicit, user provided event name, not generic one (`@timed` attribute value).